PrestaShop/PrestaShop
# [*] Project : Add customization price and weight impact, separate products in cart by customizations

## Description

Add customization price and weight impact similarly to combinations
Separate products in cart by customizations as well as combinations so that each customization can have its own line and thus quantity, price and weight
This allows modules to easily add extra data to the cart and influence the cost and shipping fees when needed
## Steps to test this Improvement
1. Apply the upgrade code if it's not a new install
   https://gist.github.com/unlocomqx/489784305d3f807404387b120f5a21ab
2. Create a text customization field on a test product
3. Save a customization in the product page and add to cart
4. Save a second customization and add to cart

The customizations will be each on a separate line (screenshots in the Forge link below)
They will be also separated in the backoffice order, invoice & emails etc

Module scénario:
5. In the table `customized_data`, add a price and weight impact to the customizations values you entered previously
This will influence the cost in cart in terms of pricing and shipping fees
1. In the table `customization_field`, set the column `is_module` to 1 to simulate a module field
   This will hide the field in the front office and make it sort of exclusive to the module which added it
   The module can then set the value of the input:hidden#product_customization_id via JS when module data is saved in the product page
2. To test the 'displayCustomization' hook, install the demo module
   https://www.dropbox.com/s/87wqjissj278ssf/customizer-demo.zip?dl=0
   Check the module ID in the `module` table , then set the column `id_module` in `customized_data` to the ID of the demo module
   Now the module can receive the value of the customization and return the desired display
   The value would be an ID of a row in the module custom table, it will then query the data, fetch a tpl for example and return the corresponding content
   Only one module will be called for a `customized_data` row and that's based on `id_module` (if not set to 0)
3. Test the reorder process after finalizing the first order

The code takes into account backward compatibility so if you have previous orders with customized data, they will be correctly displayed
The code also adds optional parameters to some core methods without creating a conflict

Module dev:
- The module creates a customization field for the product in question and sets `is_module` to 1
- It displays its interface in the product page and when the client saves the data, the module saves the customization values including `id_module`, `price` and `weight`
- The module hooks to 'displayCustomization' and displays its own data in cart, order, invoice etc
## Forge Ticket

Forge: http://forge.prestashop.com/browse/BOOM-535
Forum: https://www.prestashop.com/forums/topic/516960-cart-customizations-in-prestashop-1
